confdgnmi/src/confd_gnmi_api_adapter.py

```diff
@@ -59,32 +63,183 @@ def get_inst(cls) -> GnmiServerAdapter:
-            # TODO delete
-            delete = []
-            notif = gnmi_pb2.Notification(timestamp=0,
-                                          prefix=self.subscription_list.prefix,
-                                          alias="/alias", update=update,
-                                          delete=delete,
-                                          atomic=False)
-
-            response = gnmi_pb2.SubscribeResponse(update=notif)
-            log.debug("<== response=%s", response)
-            return response
-
-    def get_subscription_handler(self) -> SubscriptionHandler:
+        def add_path_for_monitoring(self, path, prefix):
+            log.debug("==>")
+            path_with_prefix_str = make_formatted_path(path, prefix)
+            self.monitored_paths.append(path_with_prefix_str)
+            log.debug("<==")
+
+        def kp_to_xpath(self, kp):
+            log.debug("==> kp=%s", kp)
+            xpath = _confd.xpath_pp_kpath(kp)
+            xpath = xpath.replace('"', '')
+            # for now, remove possible prefix
+            starts_slash = xpath.startswith('/')
+            xplist = xpath.split(':', 1)
+            if len(xplist) == 2:
+                xpath = xplist[1]
+                if starts_slash:
+                    xpath = '/' + xpath
+            log.debug("<== xpath=%s", xpath)
+            return xpath
+
+        def process_subscription(self, sub_sock, sub_point):
+            log.debug("==>")
+
+            def iter(kp, op, oldv, newv, state):
+                log.debug("==> kp=%s, op=%r, oldv=%s, newv=%s, state=%r", kp,
+                          op, oldv, newv, state)
+                if op == _confd.MOP_CREATED:
+                    log.debug("_confd.MOP_CREATED")
+                    # TODO CREATE not handled for now
+                if op == _confd.MOP_VALUE_SET:
+                    log.debug("_confd.MOP_VALUE_SET")
+                    self.change_db_lock.acquire()
+                    xpath = self.kp_to_xpath(kp)
+                    self.change_db.append((xpath, str(newv)))
+                    self.change_db_lock.release()
+                    # TODO MOP_VALUE_SET implement
+                elif op == _confd.MOP_DELETED:
+                    log.debug("_confd.MOP_DELETED")
+                    # TODO DELETE not handled for now
+                elif op == _confd.MOP_MODIFIED:
+                    log.debug("_confd.MOP_MODIFIED")
+                    # TODO MODIFIED not handled for now
+                else:
+                    log.warning(
+                        "Operation op=%d is not expected, kp=%s. Skipping!"
+                        , op, kp)
+                return _confd.ITER_RECURSE
+
+            cdb.diff_iterate(sub_sock, sub_point, iter, 0, None)
+            self.put_event(self.SubscriptionEvent.SEND_CHANGES)
+            log.debug("self.change_db=%s", self.change_db)
+            log.debug("<==")
+
+        def process_changes(self):
```

Comment:
This (and `process_subscription` above) definitely deserves a standalone class - it is a self-contained functionality running in a separate thread.

Reply:
Is ti some kind of pattern that separate thread must be in standalone class? I like it there, as it is related to ConfD subscription, and is part of GnmiConfDApiServerAdapter.SubscriptionHandler.

Reply:
I guess it is similar to the Handler nesting - the fact that the two functionalities are related does not mean you have to keep the code at one place, in particular if it means making the code more complex.

Reply:
Except for ;get_sample` everything is related to ConfD Subscription.

Reply:
WONT FIX for now